hamp_radar/serde_flight.py

```diff
@@ -38,66 +39,126 @@ def ppar_for_dataset(
+            - The second list contains the remaining blocks after the first
+    """
+    data = []
+    while remaining_blocks and remaining_blocks[0].mainblock.subblocks[0].tag != "PPAR":
+        data.append(remaining_blocks.pop(0))
+    return data, remaining_blocks
 
 
 def convert_to_datasetgeometries(
     pfgs: Iterable[PdsFileGeometry],
-) -> List[DatasetGeometry]:
+) -> Iterable[DatasetGeometry]:
     """
-    Converts an iterable of PdsFileGeometry instances into a list of DatasetGeometry instances.
+    Converts an iterable of PdsFileGeometry instances into an iterable of DatasetGeometry instances.
 
     This function converts a collection of PdsFileGeometry instances into
-    DatasetGeometry instances. It identifies "PPAR" tags to start new datasets and appends
-    data blocks to the most recent dataset. If consecutibe PPAR tags are found withou
-    data in between, the ppar of the current dataset is updated to match the last given PPAR.
+    DatasetGeometry instances. It identifies the next "PPAR" tag in the remaining mainblocks from
+    the PdsFileGeometry instances and the associated data thereafter in order to return the
+    geometry of the next dataset from iterating over PdsFileGeometry instances.
 
     Args:
         pfgs (Iterable[PdsFileGeometry]): An iterable of PdsFileGeometry instances, e.g. for a flight.
 
     Returns:
-        List[DatasetGeometry]: A list of DatasetGeometry instances, e.g. for each dataset in a flight.
+        Iterable[DatasetGeometry]: An iterable list of DatasetGeometry instances, e.g. for each dataset in a flight.
     """
-    import warnings
 
-    datasets = []
-    is_ppar = False
-    for pfg in pfgs:
-        for mmbg in pfg.mainblocks:
-            if mmbg.subblocks[0].tag == "PPAR":
-                ppar = ppar_for_dataset(pfg.filename, mmbg)
-                if is_ppar:
-                    # update ppar of existing dataset
-                    datasets[-1].ppar = ppar
-                else:
-                    # start new dataset
-                    datasets.append(DatasetGeometry(ppar, []))
-                    is_ppar = True
-            else:
-                if datasets == []:
-                    # skip data if no PPAR has yet been found in pfg.mainblocks
-                    warnings.warn(
-                        "Warning: data found before first PPAR is being skipped",
-                        UserWarning,
-                    )
-                else:
-                    # append to latest dataset
-                    is_ppar = False
-                    datablock = DatasetBlockGeometry(pfg.filename, mmbg)
-                    datasets[-1].data.append(datablock)
-    return datasets
+    remaining_blocks = list(generate_dbgs(pfgs))
+    while remaining_blocks:
+        ppar, remaining_blocks = next_pparblock(remaining_blocks)
+        data, remaining_blocks = next_datablocks(remaining_blocks)
+        yield DatasetGeometry(ppar, data)
```

Comment:
I'm not super sure, if this rewrite really simplifies the code. I'd be expecting that the iteration could just go once forward along the blocks and geometries would be yielded on the way. But now they have to be lists, because some jumping back and forth goes on. I could probably have a deeper look tomorrow. Currently, I think maybe something like the [peekable interator](https://more-itertools.readthedocs.io/en/stable/api.html#more_itertools.peekable) might help along.

Reply:
oh a peekable iterator is exactly what I was missing thank you! I'll edit the functions to use this and then when you have a look maybe you'll like it better. This is the best I could come up with and I prefer it to the previous function but I guess you've probably got better ideas...

Reply:
okay I had a go, what d'you think?

Reply:
It's much better, I'd say.

Reply:
I'd hope so :)